pkg/platform/runtime/artifactcache/artifactcache.go

```diff
@@ -0,0 +1,172 @@
+		}
+		err = json.Unmarshal(data, &artifacts)
+		if err != nil {
+			return nil, errs.Wrap(err, "Unable to read cached artifacts from "+infoJson)
+		}
+	}
+
+	var currentSize int64 = 0
+	artifactMap := map[artifact.ArtifactID]*cachedArtifact{}
+	for _, artifact := range artifacts {
+		currentSize += artifact.Size
+		artifactMap[artifact.Id] = &cachedArtifact{artifact.Id, artifact.ArchivePath, artifact.Size, artifact.LastAccessTime}
+	}
+
+	logging.Debug("Opened artifact cache at '%s' containing %d artifacts occupying %.1f/%.1f MB", dir, len(artifactMap), float64(currentSize)/float64(MB), float64(maxSize)/float64(MB))
+	return &ArtifactCache{dir, infoJson, maxSize, currentSize, artifactMap, sync.Mutex{}}, nil
+}
+
+// Get returns the path to the cached artifact with the given id along with true if it exists.
+// Otherwise returns an empty string and false.
+// Updates the access timestamp if possible so that this artifact is not removed anytime soon.
+func (cache *ArtifactCache) Get(a artifact.ArtifactID) (string, bool) {
+	cache.mutex.Lock()
+	defer cache.mutex.Unlock()
+
+	if artifact, found := cache.artifacts[a]; found {
+		logging.Debug("Fetched artifact '%s' as '%s'; updating access time", string(a), artifact.ArchivePath)
+		artifact.LastAccessTime = time.Now().Unix()
+		return artifact.ArchivePath, true
+	}
+	return "", false
+}
+
+// Stores the given artifact in the cache.
+// If the cache is too small, removes the least-recently accessed artifacts to make room.
+func (cache *ArtifactCache) Store(a artifact.ArtifactID, archivePath string) error {
+	cache.mutex.Lock()
+	defer cache.mutex.Unlock()
+
+	stat, err := os.Stat(archivePath)
+	if err != nil {
+		return errs.Wrap(err, "Unable to stat artifact '%s'. Does it exist?", archivePath)
+	}
+	size := stat.Size()
+
+	if size > cache.maxSize {
+		logging.Debug("Cannot avoid exceeding cache size; not storing artifact")
+		return nil
+	}
+
+	for cache.currentSize+size > cache.maxSize {
+		logging.Debug("Storing artifact in cache would exceed cache size; finding least-recently accessed artifact")
+		var lastAccessed *cachedArtifact
+		for _, artifact := range cache.artifacts {
+			if lastAccessed == nil || artifact.LastAccessTime < lastAccessed.LastAccessTime {
+				lastAccessed = artifact
+			}
+		}
+
+		if lastAccessed == nil {
+			logging.Debug("Cannot avoid exceeding cache size; not storing artifact")
+			return nil // avoid infinite loop, but this really shouldn't happen...
+		}
+
+		logging.Debug("Removing artifact '%s' last accessed on %s", lastAccessed.ArchivePath, time.Unix(lastAccessed.LastAccessTime, 0).Format(time.UnixDate))
+		err := os.Remove(lastAccessed.ArchivePath)
+		if err != nil {
+			return errs.Wrap(err, "Unable to remove cached artifact '%s'", lastAccessed.ArchivePath)
+		}
+		delete(cache.artifacts, lastAccessed.Id)
+		cache.currentSize -= lastAccessed.Size
+	}
+
+	targetPath := filepath.Join(cache.dir, string(a))
+	err = fileutils.CopyFile(archivePath, targetPath)
```

Comment:
Let's time our CopyFile operations since they could add significant time to the total setup process. Then when we reach Save() we add up the times and log to rollbar if the total time is over 5 seconds.